### PR TITLE
dyno: initial support for domains

### PR DESCRIPTION
**Pull request by benharsh**
This PR adds some basic support for rectangular and associative domains to the dyno type resolver. Some of the initial features include:
- type constructors for rectangular and associative domains
- using generic and concrete domain type expressions in function formals
- support for several built-in methods, like ``domain.idxType`` or ``domain.rank``
- calling methods and invoking iterators on the ``_domain`` record

Implementation:

The initial strategy for resolving domains relies on treating domains as a special kind of ``CompositeType``. This ``CompositeType`` will be implemented with the ``_domain`` record to support method and iterator calls. With this strategy, we should never confuse a record with a domain, which will hopefully simplify and clarify our implementation of domain behaviors.

``DomainType`` currently stores its type-constructor parameters in its ``SubstitutionsMap``, rather than having actual fields in the ``_domain`` record. This choice has been made for compatibility with the historical implementation of domains, but could be revisited in the future. ``DomainType`` also stores a ``Kind``, indicating whether it is a rectangular or associative array. Subdomains, sparse subdomains, and slices will be implemented at a future time.

The type constructor for domains in dyno is implemented by a new, special ``static_type`` method on the ``_domain`` record. If in the future our implementation of compiler-generated methods can support default values for formals, we can replace the ``static_type`` methods.

Currently the actual ``ChapelDomain`` module is beyond dyno's ability to type-resolve. For testing purposes, a very, very rudimentary ``ChapelDomain`` module has been written to demonstrate some of the functional behavior. As dyno's capabilities improve, we will eventually be able to use the proper ``ChapelDomain`` module.

TODO:
- Distributions
- Domain Literals
- RuntimeTypes

Testing:
- [x] full local
- [x] make test-dyno

### CODE REVIEW

**Comment by benharsh** [comment]
For those interested, this branch is branched from my PR #22095. You can compare the domain-specific changes here: https://github.com/benharsh/chapel/compare/dyno-generics-fixes...benharsh:chapel:dyno-domains

File: frontend/test/resolution/testDomains.cpp [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+static std::string DomainModule =
+R"""(
+module ChapelDomain {
+  record _domain {
+    var _pid: int;
+    var _instance;
+    var _unowned:bool;
+  }
+
+  proc type _domain.static_type(param rank : int, type idxType=int, param stridable: bool = false) type {
```
Comment (dlongnecke-cray) [inline frontend/test/resolution/testDomains.cpp:40]: Is `static_type` intended to be user facing? I feel weird about the `_` case. Or are we just exposing it so we can test domain features?

**Review by dlongnecke-cray — APPROVED** [review]

File: frontend/include/chpl/resolution/resolution-types.h [hunk 2]
```diff
@@ -444,7 +445,7 @@ class CallInfo {
 
 
   /** return the name of the called thing */
-  UniqueString name() const { return name_; }
+  const UniqueString name() const { return name_; }
```
[the remark this answers is not in the document]
Comment (benharsh) [reply frontend/include/chpl/resolution/resolution-types.h:448]: I figured "why not"? I think it more clearly emphasizes that the string can't be modified.

File: frontend/test/resolution/testDomains.cpp (hunk 1 above)
Comment (benharsh) [reply frontend/test/resolution/testDomains.cpp:40]: This is for domains only.

**Pull request merged**